File: Cpp/1824_minimum_sideway_jumps.cpp
// https://leetcode.cn/problems/minimum-sideway-jumps/

class Solution {
public:
    int minSideJumps(vector<int>& obstacles) {
        int n = obstacles.size();
        vector<int> dp_pre(3, 1);
        dp_pre[1] = 0;
        for (int i = 1; i < n; ++i) {
            vector<int> dp(3, n);
            int obs = obstacles[i];
            for (int lane = 0; lane < 3; ++lane) {
                if (lane == obs - 1) continue;
                for (int other = 0; other < 3; ++other) {
                    if ((other == lane) && (dp_pre[other] < n)) {
                        dp[lane] = dp_pre[lane];
                    }
                    else if ((other == obs - 1) && (lane == obstacles[i-1] - 1)) {
                        dp[lane] = min(dp[lane], dp_pre[other] + 2);
                    }
                    else {
                        dp[lane] = min(dp[lane], dp_pre[other] + 1);
                    }
                }
            }
            for (int lane = 0; lane < 3; ++lane) {
                dp_pre[lane] = dp[lane];
            }
        }
        return *min_element(dp_pre.begin(), dp_pre.end());
    }
};

File: Python3/1824_minimum_sideway_jumps.py
# https://leetcode.cn/problems/minimum-sideway-jumps/

class Solution:
    def minSideJumps(self, obstacles):
        """
        Inputs:
            obstacles: list[int]
        Outputs:
            res: int
        """
        n = len(obstacles)
        dp_pre = [1,0,1]
        for i in range(1, n):
            dp = [n,n,n]
            obs = obstacles[i]
            for lane in range(3):
                if obs == lane + 1:
                    continue
                for other in range(3):
                    if other == lane and dp_pre[lane] < n:
                        dp[lane] = dp_pre[lane]
                    elif obs == other + 1 and obstacles[i-1] == lane + 1:
                        dp[lane] = min(dp[lane], dp_pre[other]+2)
                    else:
                        dp[lane] = min(dp[lane], dp_pre[other]+1)
            for lane in range(3):
                dp_pre[lane] = dp[lane]
        return min(dp_pre)